makefiles/mcuboot.mk

```diff
@@ -1,49 +1,52 @@
+.PHONY: mcuboot mcuboot-create-key mcuboot-flash-bootloader mcuboot-flash
+
 ifdef MCUBOOT_SLOT0_SIZE
 
 IMGTOOL ?= $(RIOTTOOLS)/mcuboot/imgtool.py
-override IMGTOOL := $(abspath $(IMGTOOL))
+IMGTOOL_KEYFLAGS ?= keygen -t rsa-2048
 
-BINFILE ?= $(BINDIR)/$(APPLICATION).bin
 SIGN_BINFILE = $(BINDIR)/signed-$(APPLICATION).bin
 MCUBOOT_KEYFILE ?= $(BINDIR)/key.pem
-MCUBOOT_BIN ?= $(BINDIR)/mcuboot.bin
-MCUBOOT_BIN_URL ?= http://download.riot-os.org/mynewt.mcuboot.bin
-MCUBOOT_BIN_MD5 ?= 0c71a0589bd3709fc2d90f07a0035ce7
 
-export IMAGE_HDR_SIZE ?= 512
+MCUBOOT_ELF ?= $(ELFFILE:.elf=.mcuboot.elf)
+MCUBOOT_BIN ?= $(MCUBOOT_ELF:.elf=.bin)
+
+MCUBOOT_LOADER_BIN ?= $(BINDIR)/mcuboot.bin
+MCUBOOT_LOADER_BIN_URL ?= http://download.riot-os.org/mynewt.mcuboot.bin
+MCUBOOT_LOADER_BIN_MD5 ?= 0c71a0589bd3709fc2d90f07a0035ce7
+
+IMAGE_HDR_SIZE ?= 512
 
-$(MCUBOOT_KEYFILE) $(MCUBOOT_BIN): $(filter clean, $(MAKECMDGOALS))
+$(MCUBOOT_KEYFILE) $(MCUBOOT_LOADER_BIN): $(filter clean, $(MAKECMDGOALS))
 
 mcuboot-create-key: $(MCUBOOT_KEYFILE)
 
-ifeq ($(BINDIR)/key.pem,$(MCUBOOT_KEYFILE))
-$(MCUBOOT_KEYFILE):
+$(BINDIR)/key.pem:
 	$(Q)mkdir -p $(BINDIR)
-	$(Q)$(IMGTOOL) keygen -k $@ -t rsa-2048
-endif
+	$(Q)$(IMGTOOL) $(IMGTOOL_KEYFLAGS) -k $@
 
-mcuboot: mcuboot-create-key link
-	@$(COLOR_ECHO)
+$(MCUBOOT_ELF): $(BASELIBS)
 	@$(COLOR_ECHO) '$(COLOR_PURPLE)Re-linking for MCUBoot at $(MCUBOOT_SLOT0_SIZE)...$(COLOR_RESET)'
-	@$(COLOR_ECHO)
 	$(Q)$(_LINK) $(LINKFLAGPREFIX)--defsym=offset="$$(($(MCUBOOT_SLOT0_SIZE) + $(IMAGE_HDR_SIZE)))" \
 	$(LINKFLAGPREFIX)--defsym=length="$$(($(MCUBOOT_SLOT1_SIZE) - $(IMAGE_HDR_SIZE)))" \
-	$(LINKFLAGPREFIX)--defsym=image_header="$(IMAGE_HDR_SIZE)" -o $(ELFFILE) && \
-	$(OBJCOPY) $(OFLAGS) -Obinary $(ELFFILE) $(BINFILE) && \
-	$(IMGTOOL) sign --key $(MCUBOOT_KEYFILE) --version $(IMAGE_VERSION) --align \
-	$(MCUBOOT_IMAGE_ALIGN) -H $(IMAGE_HDR_SIZE) $(BINFILE) $(SIGN_BINFILE)
-	@$(COLOR_ECHO)
-	@$(COLOR_ECHO) '$(COLOR_PURPLE)Signed with $(MCUBOOT_KEYFILE) for version $(IMAGE_VERSION)\
-	$(COLOR_RESET)'
-	@$(COLOR_ECHO)
-
-$(MCUBOOT_BIN):
-	$(Q)$(DLCACHE) $(MCUBOOT_BIN_URL) $(MCUBOOT_BIN_MD5) $@
-
-.PHONY: mcuboot-flash-bootloader mcuboot-flash
-
-mcuboot-flash-bootloader: HEXFILE = $(MCUBOOT_BIN)
-mcuboot-flash-bootloader: $(MCUBOOT_BIN) $(FLASHDEPS)
+	$(LINKFLAGPREFIX)--defsym=image_header="$(IMAGE_HDR_SIZE)" -o $@
+
+$(SIGN_BINFILE): $(MCUBOOT_BIN) $(MCUBOOT_KEYFILE)
+	@$(COLOR_ECHO) '$(COLOR_PURPLE)Signing with $(MCUBOOT_KEYFILE) for version $(IMAGE_VERSION)$(COLOR_RESET)'
+	$(Q)$(IMGTOOL) sign --key $(MCUBOOT_KEYFILE) --version $(IMAGE_VERSION) --align \
+	$(MCUBOOT_IMAGE_ALIGN) -H $(IMAGE_HDR_SIZE) $(MCUBOOT_BIN) $(SIGN_BINFILE)
+
+ifeq ($(BUILD_IN_DOCKER),1)
+mcuboot: ..in-docker-container
```

Comment:
This currently only works in the `tests/mcuboot` application as it will not pass `mcuboot` to the docker make command.
A solution for this is to add `DOCKER_MAKECMDGOALS_POSSIBLE += mcuboot`.
However, `IMAGE_VERSION` must be set in the Makefile and not command line.
So could add a test for this to at least not try to compile.

I will find something for this.